url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/parsertables/FirstSets.java

```diff
@@ -0,0 +1,166 @@
+	all scanned nonterminals go to FIRST. This implies that one must
+	collect FIRST sets recursively, for every scanned nonterminal.
+
+	@author (c) 2000, Fritz Ritzberger
+*/
+
+class FirstSets extends UniqueAggregatingHashtable
+{
+	/**
+		Calculate Map of FIRST sets of all nonterminals contained in syntax.
+		The FIRST sets will then be lists within this Map that provides the
+		nonterminal as key.
+		@param syntax the grammar to scan.
+		@param nullAble Map containing Boolean true or false for nullability of every nonterminal.
+		@param nonterminals pre-built list of nonterminals
+	*/
+	public FirstSets(Syntax syntax, Nullable nullAble, List nonterminals)
+		throws ParserBuildException
+	{
+		// collect FIRST sets into this Hashtable where each
+		// nonterminal has a List of FIRST-symbols
+		Map done = new Hashtable(nonterminals.size());	// avoid recursion
+		for (int i = 0; i < nonterminals.size(); i++)	{
+			String nonterm = (String) nonterminals.get(i);
+			generateFirstSet(syntax, nullAble, nonterm, done);
+		}
+	}
+
+	private void generateFirstSet(Syntax syntax, Nullable nullAble, String nonterminal, Map done)
+		throws ParserBuildException
+	{
+		if (get(nonterminal) != null)
+			return;	// alreay done
+		
+		if (done.get(nonterminal) != null)
+			return;
+		done.put(nonterminal, nonterminal);	// avoid endless loops
+		//System.err.println("generateFirstSet("+nonterminal+")");
+
+		// The FIRST set of a nonterminal are all symbols
+		// that are on first position of the right side of all rules
+		// where the nonterminal is on the left side.
+		for (int k = 0; k < syntax.size(); k++)	{
+			Rule rule = syntax.getRule(k);
+			String nonterm = rule.getNonterminal();	// left side of derivation
+
+			if (nonterminal.equals(nonterm))	{	// this rule derives the nonterminal
+				// if left side is empty, add NULL to FIRST of nonterminal
+				if (rule.rightSize() <= 0)	{
+					put(nonterm, Nullable.NULL);
+				}
+				else	{	// there are symbols on left side
+					boolean nullable = true;	// enable loop until nullable
+	
+					// While nullable, add the symbol, shift to the next and
+					// check if it is nullable again.
+					for (int i = 0; nullable && i < rule.rightSize(); i++)	{
+						String symbol = rule.getRightSymbol(i);
+						
+						nullable = false;	// assume it is a terminal
+
+						if (Token.isTerminal(symbol))	{
+							put(nonterm, symbol);
+						}
+						else	{
+							// If there is a nonterminal on first position, add its FIRST set
+							// FIRST set of this nonterminal, but without null-word.
+							try	{
+								generateFirstSet(syntax, nullAble, symbol, done);	// enter recursion
+
+								List list = (List) get(symbol);	// get the results
+								for (int j = 0; list != null && j < list.size(); j++)	{
+									String s = (String) list.get(j);
+									put(nonterm, s);
+								}
+
+								nullable = nullAble.isNullable(symbol);
+							}
+							catch (Exception ex)	{
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)